dask_ml/model_selection/_hyperband.py

```diff
@@ -0,0 +1,597 @@
+    2. The testing dataset should fit in the memory of a single worker
+
+    References
+    ----------
+    .. [1] "Hyperband: A novel bandit-based approach to hyperparameter
+           optimization", 2016 by L. Li, K. Jamieson, G. DeSalvo, A.
+           Rostamizadeh, and A. Talwalkar.  https://arxiv.org/abs/1603.06560
+
+    """
+
+    def __init__(
+        self,
+        estimator,
+        param_distribution,
+        max_iter=81,
+        aggressiveness=3,
+        test_size=None,
+        patience=False,
+        tol=1e-3,
+        random_state=None,
+        scoring=None,
+    ):
+        self.estimator = estimator
+        self.param_distribution = param_distribution
+        self.max_iter = max_iter
+        self.aggressiveness = aggressiveness
+
+        super(HyperbandSearchCV, self).__init__(
+            estimator,
+            param_distribution,
+            max_iter=max_iter,
+            patience=patience,
+            tol=tol,
+            test_size=test_size,
+            random_state=random_state,
+            scoring=scoring,
+        )
+
+    def _get_SHAs(self, brackets):
+
+        patience = _get_patience(self.patience, self.max_iter, self.aggressiveness)
+
+        # This is the first time self.random_state is used after
+        # HyperbandSearchCV.fit is called.
+        seed_start = check_random_state(self.random_state).randint(2 ** 31)
+        self._SHA_seed = seed_start
+
+        # These brackets are ordered by adaptivity; bracket=0 is least adaptive
+        SHAs = [
+            (
+                b,
+                SuccessiveHalvingSearchCV(
+                    self.estimator,
+                    self.param_distribution,
+                    n_initial_parameters=n,
+                    start_iter=r,
+                    adaptive_max_iter=b + 1,
+                    aggressiveness=self.aggressiveness,
+                    max_iter=self.max_iter,
+                    patience=patience,
+                    tol=self.tol,
+                    test_size=self.test_size,
+                    random_state=seed_start + b if b != 0 else self.random_state,
+                    scoring=self.scoring,
+                ),
+            )
+            for b, (n, r) in brackets.items()
+        ]
+        return SHAs
+
+    @gen.coroutine
+    def _fit(self, X, y, **fit_params):
+        X = self._check_array(X)
+        y = self._check_array(y, ensure_2d=False)
+        scorer = check_scoring(self.estimator, scoring=self.scoring)
+
+        brackets = _get_hyperband_params(self.max_iter, eta=self.aggressiveness)
+
+        SHAs = self._get_SHAs(brackets)
+        # Which bracket to run first? Going to go with most adaptive;
+        # hopefully less adaptive can fill in for any blank spots
+        #
+        # _brackets is ordered from largest to smallest
+        _brackets_ids = [b for b, SHA in SHAs]
+        _SHAs = yield [SHA.fit(X, y, **fit_params) for b, SHA in SHAs]
```

Comment:
Do we want all the SHAs fit in parallel? I may be wrong, but this seems to be sequential. I don't think the `yield` is doing anything.

Each SHA.fit returns a concrete SucessiveHalvingSearchCV, not a future, so we make a list of concrete SucessiveHalvingSearchCV objects and yield that.

Reply:
Surprisingly, the list can be `yield`ed directly because it's in a function decorated with `gen.coroutine`:

> In decorated coroutines, it is possible to yield the list or dict directly:
> ``` python
> @gen.coroutine
> def parallel_fetch_decorated(url1, url2):
>     resp1, resp2 = yield [http_client.fetch(url1),
>                           http_client.fetch(url2)]
> ```
> —https://www.tornadoweb.org/en/stable/guide/coroutines.html#parallelism

Reply:
I believe in that example the function called in the list comprehension is a coroutine. Here it’s a regular method, which itself calls a coroutine. Does that matter? I’m not at a computer to check.

Reply:
> Does that matter?

I don't think so, at least not if the function `http_client.fetch` has internal calls to Tornado. Here's an example of what I mean: https://gist.github.com/stsievert/0e080e8b49bd15382972172e97252315

Either way,  I've added calling the coroutine more explicit with `yield [SHA._fit(...) for SHA in SHAs]`.

Functionally, I'm not sure what difference it makes: the data I've recorded indicate the brackets are run in parallel.

<img width="419" alt="Screen Shot 2019-04-15 at 8 42 53 AM" src="https://user-images.githubusercontent.com/1320475/56137448-6832eb00-5f84-11e9-9d9a-8d992abec379.png">